File: usage-daemon/src/providers/context7.js
// Context7 usage provider plugin — descriptor interface.
//
// Context7's dashboard stats endpoint
//   GET https://context7.com/api/dashboard/stats/<teamId>
//     -> { success: true, data: { quotaLimit, userRequests, ownerPlan,
//                                  creditBalance, dailyStats: [...] } }
// gives the monthly request quota (quotaLimit) vs requests consumed this
// cycle (userRequests). There is no reset-date field in the payload — the
// window's resets_at is left null.
//
// Auth: browser session cookie for context7.com (cookie kind, same as
// tavily/mistral/ollama). Context7 uses Clerk for auth — its __session JWT
// is extremely short-lived (~60s), so a hand-pasted cookie goes stale almost
// immediately. Use cookie_from_firefox = "context7.com" so the daemon pulls
// a fresh cookie out of Firefox before each poll instead of relying on a
// static paste.
//
// parse() is a PURE function of the envelope JSON text.

export const REQUESTS_COLOR = '#F0E442'; // Okabe-Ito yellow

export const ID = 'context7';
export const LABEL = 'Context7';

const API_URL = 'https://context7.com';
const USER_AGENT = 'usage-daemon/0.1';

export class AuthExpiredError extends Error {
  constructor(msg = 'Context7 session cookie missing or expired') {
    super(msg);
    this.code = 'auth_expired';
  }
}

export class RateLimitedError extends Error {
  constructor(retryAfter = null) {
    super('rate_limited');
    this.code = 'rate_limited';
    this.retryAfter = retryAfter;
  }
}

function clampPct(n) {
  if (typeof n !== 'number' || !Number.isFinite(n)) return null;
  return Math.max(0, Math.min(100, n));
}

// Pure function of the envelope JSON text — no fs/network.
// envelope: { quotaLimit, userRequests, ownerPlan, creditBalance }
// (the dashboard stats endpoint's `data` object)
export function parse(raw) {
  let env;
  try {
    env = typeof raw === 'string' ? JSON.parse(raw) : raw;
  } catch {
    throw new AuthExpiredError('unparseable context7 envelope');
  }
  if (!env || typeof env !== 'object') {
    throw new AuthExpiredError('unparseable context7 envelope');
  }
  if (env.success !== true || !env.data) {
    throw new AuthExpiredError('context7: not authenticated');
  }

  const data = env.data;
  const consumed = typeof data.userRequests === 'number' ? data.userRequests : null;
  const cap = typeof data.quotaLimit === 'number' ? data.quotaLimit : null;

  if (consumed == null) {
    throw new AuthExpiredError('no usable Context7 request figure in envelope');
  }

  const windows = [
    {
      id: 'requests',
      label: 'Requests/mo',
      letter: 'Rq',
      pct: cap && cap > 0 ? clampPct((100 * consumed) / cap) : null,
      used: cap != null ? Math.max(0, cap - consumed) : null, // remaining, not consumed
      used_is_remaining: true,
      cap,
      unit: 'requests',
      resets_at: null,
      color: REQUESTS_COLOR,
      will_deplete: false,
    },
  ];

  return {
    tier: data.ownerPlan ?? null,
    windows,
    segments: [],
    // carried out of parse so meta() can surface the raw figures too.
    _context7: {
      quotaLimit: cap,
      userRequests: consumed,
      ownerPlan: data.ownerPlan ?? null,
      creditBalance: typeof data.creditBalance === 'number' ? data.creditBalance : null,
    },
  };
}

function createProvider() {
  let cookie = null;
  let teamId = null;
  let lastStats = null; // surfaced via meta()

  return {
    id: ID,
    label: LABEL,
    auth: { kind: 'cookie' },

    config() {
      return {
        id: ID,
        label: LABEL,
        usageUrl: 'https://context7.com/dashboard',
        auth: { kind: 'cookie' },
        // Support service (docs/API access), not an AI plan: a metered API
        // that backs the work rather than being the work. Clients render
        // these compactly instead of giving them a full plan card.
        category: 'support',
        windows: [{ id: 'requests', label: 'Requests/mo', color: REQUESTS_COLOR }],
        tiers: [],
      };
    },

    configure(cfg = {}) {
      // !== undefined so configure({cookie:''}) can explicitly clear it.
      if (cfg.cookie !== undefined) cookie = cfg.cookie ? String(cfg.cookie).trim() : null;
      if (cfg.team_id !== undefined) teamId = cfg.team_id ? String(cfg.team_id).trim() : null;
    },

    async fetch() {
      if (!cookie) throw new AuthExpiredError('no Context7 session cookie configured');
      if (!teamId) throw new Error('context7: no team_id configured (see config.example.toml)');

      const res = await fetch(`${API_URL}/api/dashboard/stats/${encodeURIComponent(teamId)}`, {
        headers: {
          Cookie: cookie,
          'User-Agent': USER_AGENT,
          Accept: 'application/json',
          Referer: 'https://context7.com/dashboard',
        },
      });

      if (res.status === 401 || res.status === 403) throw new AuthExpiredError();
      if (res.status === 429) {
        throw new RateLimitedError(Number(res.headers.get('retry-after')) || null);
      }
      if (!res.ok) throw new Error(`context7.com HTTP ${res.status}`);

      const body = await res.json();
      // Cache figures for meta() (parse() computes them too, but the runner
      // calls meta() independently of parse()).
      lastStats = parse(JSON.stringify(body))._context7;
      return JSON.stringify(body);
    },

    intervalSeconds() {
      return 300;
    },

    meta() {
      return lastStats
        ? {
            quota_limit: lastStats.quotaLimit,
            user_requests: lastStats.userRequests,
            owner_plan: lastStats.ownerPlan,
            credit_balance: lastStats.creditBalance,
          }
        : {};
    },

    parse,
  };
}

export { createProvider };

File: usage-daemon-v3/usage_daemon/providers/context7.py
"""Context7 usage provider plugin (port of src/providers/context7.js, plus a
Clerk session-refresh step).

Context7 authenticates the dashboard with Clerk, which issues its
``__session`` JWT with only a ~60-second lifetime and re-issues it while the
dashboard tab is open. A cookie pulled from Firefox on a poll schedule is
therefore almost always expired, so ``fetch()`` mints a fresh session JWT via
Clerk's frontend token endpoint (``POST /v1/client/sessions/<sid>/tokens``,
the same call Clerk's SDK makes) before hitting the stats API.
"""

from __future__ import annotations

import base64
import json
import math
from typing import Any
from urllib.parse import quote

import httpx

from ..errors import AuthExpiredError, RateLimitedError
from ..httputil import create_client

REQUESTS_COLOR = "#F0E442"

ID = "context7"
LABEL = "Context7"

API_URL = "https://context7.com"
FRONTEND_URL = "https://clerk.context7.com"
USER_AGENT = "usage-daemon/0.1"


def _jwt_payload(jwt: str) -> dict | None:
    try:
        seg = jwt.split(".")[1]
        seg += "=" * (-len(seg) % 4)
        data = json.loads(base64.urlsafe_b64decode(seg))
    except Exception:
        return None
    return data if isinstance(data, dict) else None


def _session_id_from_jwt(jwt: str) -> str | None:
    payload = _jwt_payload(jwt)
    sid = payload.get("sid") if payload else None
    return sid if isinstance(sid, str) and sid else None


def _session_jwt_from_cookie_header(header: str) -> str | None:
    for part in header.split(";"):
        part = part.strip()
        if not part.lower().startswith("__session="):
            continue
        value = part.split("=", 1)[1].strip().strip('"')
        return value or None
    return None


async def _clerk_refresh_jwt(cookie_header: str, client: httpx.AsyncClient) -> str | None:
    """Mint a fresh Clerk session JWT from the browser's session cookies.

    Clerk's ``__session`` JWT lives ~60s, so we exchange the stored cookies
    (which identify the live session) for a brand-new JWT, then use it as the
    Bearer token against the Context7 API. Returns None when the cookie set
    has no usable Clerk session.
    """
    session_jwt = _session_jwt_from_cookie_header(cookie_header)
    if not session_jwt:
        return None
    sid = _session_id_from_jwt(session_jwt)
    if not sid:
        return None
    try:
        res = await client.post(
            f"{FRONTEND_URL}/v1/client/sessions/{sid}/tokens",
            headers={
                "Origin": "https://context7.com",
                "Referer": "https://context7.com/dashboard",
                "Cookie": cookie_header,
                "Content-Type": "application/json",
                "User-Agent": USER_AGENT,
            },
            json={},
        )
        if res.status_code != 200:
            return None
        jwt = res.json().get("jwt")
    except Exception:
        return None
    return jwt if isinstance(jwt, str) and jwt else None


def _clamp_pct(n):
    if not isinstance(n, (int, float)) or isinstance(n, bool) or not math.isfinite(n):
        return None
    return max(0.0, min(100.0, float(n)))


def parse(raw) -> dict:
    try:
        env = json.loads(raw) if isinstance(raw, str) else raw
    except Exception:
        raise AuthExpiredError("unparseable context7 envelope")
    if not isinstance(env, dict):
        raise AuthExpiredError("unparseable context7 envelope")
    if env.get("success") is not True or not isinstance(env.get("data"), dict):
        raise AuthExpiredError("context7: not authenticated")

    data = env["data"]
    consumed = data.get("userRequests") if isinstance(data.get("userRequests"), (int, float)) and not isinstance(data.get("userRequests"), bool) else None
    cap = data.get("quotaLimit") if isinstance(data.get("quotaLimit"), (int, float)) and not isinstance(data.get("quotaLimit"), bool) else None
    if consumed is None:
        raise AuthExpiredError("no usable Context7 request figure in envelope")

    windows = [{
        "id": "requests",
        "label": "Requests/mo",
        "letter": "Rq",
        "pct": _clamp_pct((100 * consumed) / cap) if cap is not None and cap > 0 else None,
        "used": max(0, cap - consumed) if cap is not None else None,
        "used_is_remaining": True,
        "cap": cap,
        "unit": "requests",
        "resets_at": None,
        "color": REQUESTS_COLOR,
        "will_deplete": False,
    }]

    return {
        "tier": data.get("ownerPlan") or None,
        "windows": windows,
        "segments": [],
        "_context7": {
            "quotaLimit": cap,
            "userRequests": consumed,
            "ownerPlan": data.get("ownerPlan") or None,
            "creditBalance": data.get("creditBalance") if isinstance(data.get("creditBalance"), (int, float)) and not isinstance(data.get("creditBalance"), bool) else None,
        },
    }


def create() -> dict:
    state: dict[str, Any] = {"cookie": None, "team_id": None, "last_stats": None}

    def config() -> dict:
        return {
            "id": ID,
            "label": LABEL,
            "usageUrl": "https://context7.com/dashboard",
            "auth": {"kind": "cookie"},
            "category": "support",
            "windows": [{"id": "requests", "label": "Requests/mo", "color": REQUESTS_COLOR}],
            "tiers": [],
        }

    def configure(cfg: dict | None = None) -> None:
        cfg = cfg or {}
        if "cookie" in cfg:
            state["cookie"] = str(cfg["cookie"]).strip() if cfg["cookie"] else None
        if "team_id" in cfg:
            state["team_id"] = str(cfg["team_id"]).strip() if cfg["team_id"] else None

    async def fetch() -> str:
        if not state["cookie"]:
            raise AuthExpiredError("no Context7 session cookie configured")
        if not state["team_id"]:
            raise RuntimeError("context7: no team_id configured (see config.example.toml)")
        c = create_client()
        try:
            jwt = await _clerk_refresh_jwt(state["cookie"], c)
            if not jwt:
                raise AuthExpiredError(
                    "context7: could not refresh Clerk session — log in to context7.com "
                    "in Firefox and refresh the cookie"
                )
            res = await c.get(
                f"{API_URL}/api/dashboard/stats/{quote(state['team_id'], safe='')}",
                headers={
                    "Authorization": f"Bearer {jwt}",
                    "User-Agent": USER_AGENT,
                    "Accept": "application/json",
                    "Referer": "https://context7.com/dashboard",
                },
            )
        finally:
            await c.aclose()
        if res.status_code in (401, 403):
            raise AuthExpiredError("context7 session logged out — log in at context7.com and refresh the cookie")
        if res.status_code == 429:
            ra = res.headers.get("retry-after")
            raise RateLimitedError(int(ra) if ra and ra.isdigit() else None)
        if res.status_code >= 400:
            raise RuntimeError(f"context7.com HTTP {res.status_code}")
        state["last_stats"] = parse(res.text)["_context7"]
        return res.text

    def meta() -> dict:
        s = state["last_stats"]
        if not s:
            return {}
        return {
            "quota_limit": s["quotaLimit"],
            "user_requests": s["userRequests"],
            "owner_plan": s["ownerPlan"],
            "credit_balance": s["creditBalance"],
        }

    return {
        "id": ID,
        "label": LABEL,
        "auth": {"kind": "cookie"},
        "config": config,
        "configure": configure,
        "fetch": fetch,
        "interval_seconds": lambda: 300,
        "meta": meta,
        "parse": parse,
    }
